src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+	D_ASSERT(p > 0);
+	for (i = p - 1, j = 0; i >= 0; i--)
+		parity_bufs[j++] = &buf[i * cell_bytes];
+
+	if (entry->ae_codec == NULL)
+		entry->ae_codec =
+		obj_ec_codec_get(daos_obj_id2class(entry->ae_oid.id_pub));
+	ec_encode_data(cell_bytes, k, p, entry->ae_codec->ec_gftbls, data,
+		       parity_bufs);
+}
+
+/* Xstream offload function for partial stripe update. Fetches the old data
+ * from the data target(s) and updates the parity.
+ */
+static void
+agg_process_partial_stripe_ult(void *arg)
+{
+	struct ec_agg_stripe_ud	*stripe_ud = (struct ec_agg_stripe_ud *)arg;
+	struct ec_agg_entry	*entry = stripe_ud->asu_agg_entry;
+	uint8_t			*bit_map = stripe_ud->asu_bit_map;
+	unsigned int		 p = entry->ae_oca->u.ec.e_p;
+	unsigned int		 cell_cnt = stripe_ud->asu_cell_cnt;
+	int			 rc = 0;
+
+	/* Fetch the data cells on other shards. For parity update,
+	 * the bitmap is set for the same cells as are replicated.
+	 */
+	rc = agg_fetch_odata_cells(entry, bit_map, cell_cnt,
+				   stripe_ud->asu_recalc);
+	if (rc)
+		goto out;
+
+	if (p > 1) {
+		rc = agg_fetch_remote_parity(entry);
+		if (rc)
+			goto out;
+	}
+
+	if (stripe_ud->asu_recalc)
+		agg_recalc_parity(entry, bit_map, cell_cnt);
+	else
+		rc = agg_update_parity(entry, bit_map, cell_cnt);
+
+out:
+	ABT_eventual_set(stripe_ud->asu_eventual, (void *)&rc, sizeof(rc));
+
+}
+
+/* Driver function for partial stripe update. Fetches the data and then invokes
+ * second function to update the parity.
+ */
+static int
+agg_process_partial_stripe(struct ec_agg_entry *entry)
+{
+	struct ec_agg_stripe_ud	 stripe_ud = { 0 };
+	struct ec_agg_extent	*extent;
+	int			*status;
+	uint8_t			*bit_map = NULL;
+	uint8_t			 fcbit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	uint8_t			 tbit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned long            ss;
+	unsigned int		 i, full_cell_cnt = 0;
+	unsigned int		 cell_cnt = 0;
+	unsigned int		 estart, elen = 0;
+	unsigned int		 eend = 0;
+	int			 rc = 0;
+
+	/* For each contiguous extent, constructable from the extent list,
+	 * determine how many full cells, and how many cells overall,
+	 * are contained in the constructed extent.
+	 */
+	ss = k * len * entry->ae_cur_stripe.as_stripenum;
+	estart = entry->ae_cur_stripe.as_offset;
+	d_list_for_each_entry(extent, &entry->ae_cur_stripe.as_dextents,
 			      ae_link) {
-		daos_epoch_range_t	 epoch_range = { 0 };
-		int			 erc = 0;
-
-		epoch_range.epr_lo = epoch_range.epr_hi = agg_extent->ae_epoch;
-
-		erc = vos_obj_array_remove(agg_param->ap_cont_handle,
-					   entry->ae_oid, &epoch_range,
-					   &entry->ae_dkey, &entry->ae_akey,
-					   &agg_extent->ae_orig_recx);
-		if (erc)
-			D_ERROR("vos_obj_array_remove failed: "DF_RC"\n",
-				DP_RC(erc));
-		if (!rc && erc)
-			rc = erc;
-		ext_counter++;
-		if (ext_counter >= ext_total)
-			break;
+		D_ASSERT(!extent->ae_hole);
+		if (extent->ae_epoch <= entry->ae_par_extent.ape_epoch)
```

Comment:
actually, should never ==, can add D_ASSERT(extent->ae_epoch != entry->ae_par_extent.ape_epoch);

Reply:
You're correct: it "should" not be equal, but I changed it to <= to fix the following issue.  If the replica was the highest epoch in prior run of EC aggregation, then the parity was written at this same epoch.  If the vos_obj_array_remove call fails (which it should not do), then equality will hold on the next EC aggregation run.  The remove will then be retried whereas an assert will stop the process.